url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/utility/geo/DataOfLine.java

```diff
@@ -0,0 +1,182 @@
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.util.ArrayList;
+
+import utility.Compare.DataOfPoint;
+import utility.security.DataIO;
+import utility.security.Point;
+import utility.security.RSA;
+import utility.security.SecurityUtility;
+
+public class DataOfLine {
+
+	public Long lid;
+	public Line line;
+	public static Long M = (long) 100000000;
+	public String signature;
+	
+	public DataOfLine() {
+		line = new Line();
+		// TODO Auto-generated constructor stub
+	}
+
+	public DataOfLine(long _lid, Line _line){
+		lid = _lid;
+		line = _line;
+		signature = signWithRSA(new RSA());
+	}
+	
+	public DataOfLine(byte[] buffer){
+		this();
+		this.loadFromBytes(buffer);
+	}
+	
+	public byte[] writeToBytes(){
+		ByteArrayOutputStream baos = new ByteArrayOutputStream();
+		DataOutputStream dos = new DataOutputStream(baos);
+		try {
+			dos.writeLong(lid);
+		} catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		line.writeToFile(dos);
+		DataIO.writeString(dos, signature);
+		try {
+			dos.flush();
+			dos.close();
+		} catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		return baos.toByteArray();
+	}
+	
+	public void loadFromBytes(byte[] buffer){
+		DataInputStream dis = new DataInputStream(new ByteArrayInputStream(buffer));
+		try {
+			lid = dis.readLong();
+		} catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		line.readFromFile(dis);
+		signature = DataIO.readString(dis);
+		try {
+			dis.close();
+		} catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+	}
+	
+	public void loadFromFile(DataInputStream dis){
+		try {
+			lid = dis.readLong();
+		} catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		line.readFromFile(dis);
+	}
+
+	public long getLineId(){
+		return lid;
+	}
+	
+	public static long calcLineId(int id1, int id2){
+		if(id1 > id2){
+			return ((long)id2) * M + id1;
+		}
+		return ((long)id1) * M + id2;
+	}
+	
+	public static long calcLineId(long id1, long id2){
+		if(id1 > id2){
+			return ((long)id2) * M + id1;
+		}
+		return id1 * M + id2;
+	}
+	
+	public String getDigest(){
+		return SecurityUtility.computeGeneralHashValue(new String[]{
+			new Integer((int)(lid / M)).toString(),
+			new Integer((int)(lid % M)).toString(),
+			line.getDigest()
+		});
+	}
+	
+	public String signWithRSA(RSA rsa){
+		return rsa.encrypt(getDigest());
+	}
+	
+	/**
+	 * @param args
+	 * @throws IOException 
+	 */
+	public static void main(String[] args) throws IOException {
+		// TODO Auto-generated method stub
+		
+		Line l = new Line(new int[]{3, 1, 4}, new int[]{6, 1, 5});
+		DataOfLine dataOfLine = new DataOfLine(1, l);
+		DataOutputStream dos;
+		try {
+			dos = new DataOutputStream(new FileOutputStream(new File("./tmp/dataofline")));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *dos*. The resource is referenced at lines 134, 136, 137. The resource is closed at line 138. There are other execution paths that do not contain closure statements, for example, when FilterOutputStream.write throws an exception. Ensure *dos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).